// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.wave";
option java_outer_classname = "WaveProto";
option objc_class_prefix = "HLW";

package wave;

// The wave service definition.
service WaveService {
  // Request wave elevation
  rpc GetElevation (Point) returns (Elevation) {}
  rpc GetElevations (Point) returns (stream Elevation) {}
}

// The request message containing the point coordinates.
message Point {
  repeated double x = 1;
  repeated double y = 2;
  double t = 3;
  double dt = 4;
  double t_start = 5;
  double t_end = 6;
}

// The response message containing the elevation
message Elevation {
  repeated double z = 1;
  double t = 2;
}


// The discrete directional wave spectrum
message FlatDiscreteDirectionalWaveSpectrum {
  double a = 1;       //!< Amplitude (in m)
  double omega = 2;   //!< Angular frequencies the spectrum was discretized at (in rad/s)
  double psi = 3;     //!< Directions between 0 & 2pi the spatial spreading was discretized at (in rad)
  double k = 4;       //!< Discretized wave number (for each frequency) in rad/m
  double phase = 5;   //!< Random phases, for each (direction,frequency) couple (but time invariant) in radian
}
